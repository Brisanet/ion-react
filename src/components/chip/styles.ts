import stitches from '../../stitches.config';

const { styled } = stitches;

export const ChipStyle = styled('button', {
  border: '1px solid $neutral4',
  borderRadius: '100px',
  cursor: 'pointer',
  outline: 'none',
  backgroundColor: '$neutral1',
  color: '$primaryColor',
  fontWeight: '600',

  '&:hover': {
    backgroundColor: '$secondaryColor',
    border: '1px solid $primary3',
    color: '$primary5',
  },

  variants: {
    size: {
      sm: {
        padding: '4px 8px',
        lineHeight: '16px',
        fontSize: '12px',
      },
      md: {
        padding: '6px 12px',
        lineHeight: '20px',
        fontSize: '14px',
      },
    },
  },
});
