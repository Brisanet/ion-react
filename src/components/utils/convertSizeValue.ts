/**
 * Get a size value based on its type.
 * @param {string | number} value - The size value.
 * @returns The value itself if it's a string, or the value in pixels if it is a number.
 **/
export const convertSizeValue = (value: string | number): string =>
  typeof value === 'number' ? `${value}px` : value;
