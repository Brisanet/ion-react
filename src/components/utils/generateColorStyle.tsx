// Tipo para objeto de estilos  da função geradora de estilos
type ColorStyle = {
    background: string;
    color: string;
    border?: string;
  };
  
  /* 
    Função que gera cores de Background, Border e cor de
    fonte a fim de reduzir repetições de codigo
  
    Seguir o padrão de parâmetros: 
    1- Cor do background;
    2- Cor da fonte do botão;
    3- (Opicional) Cor da borda;
  */
 export function generateColorStyle(
    bgColor: string,
    fontColor: string,
    borderColor: string
  ) {
    let styles: ColorStyle = {
      background: bgColor,
      color: fontColor,
    };
  
    borderColor && (styles.border = `1px solid ${borderColor}`);
  
    return styles;
  }