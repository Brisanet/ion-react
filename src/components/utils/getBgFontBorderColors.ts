interface ObjStyles {
  background: string;
  color: string;
  border?: string;
}

/**
 * Function to prevent repetition of style blocks.
 *
 * @param background - Background color
 * @param color - The font color
 * @param border - Border attribute styles - example: ´1px solid blue´
 * @returns A Javascript object containing color properties.
 */
export function getBgFontBorderColors({
  background,
  color,
  border = '',
}: ObjStyles): ObjStyles {
  return {
    background,
    color,
    border,
  };
}
