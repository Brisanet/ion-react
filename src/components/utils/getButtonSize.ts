/**
 * Function to prevent repetition of style blocks.
 *
 * @param buttonSize - Height and min-width
 * @param padding - Button padding
 * @param fontSize - The size of font
 * @returns An Javascript object containing size properties.
 */
export function getButtonSize(
  buttonSize: string,
  padding: string,
  fontSize: string
) {
  let size = {
    height: buttonSize,
    minWidth: buttonSize,
    padding: padding,
    fontSize: fontSize,
  };

  return size;
}
