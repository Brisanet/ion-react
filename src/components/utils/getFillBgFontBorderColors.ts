type ColorStyleResponse = {
  border: string;
  backgroundColor: string;
  color: string;
  svg: {
    fill: string;
  };
};

interface StyleProps {
  svgFillColor: string;
  backgroundColor: string;
  fontColor: string;
  borderColor: string;
}

/**
 * Function to prevent repetition of style blocks.
 *
 * @param svgFillColor - Fill color to be used in SVG
 * @param backgroundColor - Background color
 * @param fontColor - The font color
 * @param borderColor - Border color
 * @returns A Javascript object containing style properties.
 */
export function getFillBgFontBorderColors({
  backgroundColor,
  fontColor,
  borderColor,
  svgFillColor,
}: StyleProps) {
  let styles: ColorStyleResponse = {
    backgroundColor: backgroundColor,
    color: fontColor,
    border: `1px solid ${borderColor}`,
    svg: {
      fill: svgFillColor,
    },
  };

  return styles;
}
