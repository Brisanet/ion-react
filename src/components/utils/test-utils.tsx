import theme from '@ion/styles/theme';
import { render, RenderResult } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

export const renderWithTheme = (children: React.ReactNode): RenderResult =>
  render(<ThemeProvider theme={theme}>{children}</ThemeProvider>);
