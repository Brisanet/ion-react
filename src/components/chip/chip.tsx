import { ChipStyle } from './styles';
import { SizeType } from '../../core/types/size';
import React from 'react';

export interface ChipProps {
  label: string;
  size?: SizeType;
  disabled?: boolean;
  selected?: boolean;
  handleClick?: () => void;
}

const IonChip = ({
  label,
  size = 'sm',
  disabled = false,
  selected = false,
  handleClick,
}: ChipProps) => {
  return (
    <ChipStyle
      size={size}
      disabled={disabled}
      selected={selected}
      onClick={handleClick}
      data-testid="ion-chip"
    >
      <span>{label}</span>
    </ChipStyle>
  );
};

export default IonChip;
