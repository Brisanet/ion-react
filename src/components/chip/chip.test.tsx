import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { renderWithTheme } from '../utils/test-utils';
import { ChipProps, ChipSizes, IonChip } from './chip';

const clickEvent = jest.fn();
const defaultChip: ChipProps = {
  label: 'Ragnarok!',
  handleClick: clickEvent,
};
const chipId = 'ion-chip';

const sut = (props: ChipProps = defaultChip) => {
  return renderWithTheme(<IonChip {...props} />);
};

describe('IonChip', () => {
  describe('Default', () => {
    beforeEach(() => {
      sut();
    });

    it('should render chip', () => {
      expect(screen.queryAllByTestId(chipId)).toHaveLength(1);
    });

    it('should render chip with correct value', () => {
      expect(screen.queryAllByText(defaultChip.label)).toHaveLength(1);
    });

    it('should render chip with correct value', async () => {
      await userEvent.click(screen.getByTestId(chipId));
      expect(clickEvent).toHaveBeenCalled();
    });

    it('should render chip not disabled by default', () => {
      expect(screen.getByTestId(chipId)).not.toBeDisabled();
    });
  });

  describe('Custom Props', () => {
    it('should render chip selected', () => {
      const { container } = sut({ ...defaultChip, selected: true });
      expect(container).toMatchSnapshot();
    });

    it.each<ChipSizes>(['sm', 'md'])('should render chip %s size', (size) => {
      const { container } = sut({ ...defaultChip, size });
      expect(container).toMatchSnapshot();
    });
  });

  describe('With Icon', () => {
    const iconName = 'pencil';

    beforeEach(() => {
      sut({ ...defaultChip, icon: iconName });
    });

    it('should render pencil icon', () => {
      expect(screen.getByTestId(`ion-icon-${iconName}`)).toBeInTheDocument();
    });

    it('should render icon sm by default', () => {
      const smSize = '16';
      expect(screen.getByTestId(`ion-icon-${iconName}`)).toHaveStyleRule(
        'height',
        `${smSize}px`
      );
    });

    it('should render icon md when chip is md', () => {
      const mdSize = '20';
      const mdIconName = 'alert';
      sut({ ...defaultChip, size: 'md', icon: mdIconName });
      expect(screen.getByTestId(`ion-icon-${mdIconName}`)).toHaveStyleRule(
        'height',
        `${mdSize}px`
      );
    });
  });

  afterEach(() => {
    clickEvent.mockClear();
  });
});
