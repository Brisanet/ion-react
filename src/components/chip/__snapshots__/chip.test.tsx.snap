// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IonChip Custom Props should render chip md size 1`] = `
.c0 {
  background-color: #fcfcfd;
  border: 1px solid #ced2db;
  border-radius: 100px;
  color: #0858ce;
  cursor: pointer;
  font-weight: 600;
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 8px;
  height: 32px;
  padding: 6px 12px;
  font-size: 1.4rem;
  line-height: 2rem;
}

.c0:focus-visible {
  outline: 2px solid #146ff5;
  outline-offset: 2px;
}

.c0 svg {
  fill: #0858ce;
}

.c0:hover,
.c0:focus-visible {
  background-color: #ebf3fe;
  border-color: #84b4fa;
  color: #146ff5;
}

.c0:hover svg,
.c0:focus-visible svg {
  fill: #146ff5;
}

.c0:active {
  background-color: #b5d2fc;
  border-color: #146ff5;
  color: #06439d;
}

.c0:active svg {
  fill: #06439d;
}

.c0:disabled {
  cursor: not-allowed;
  background-color: #e4e6eb;
  border-color: #aeb2bd;
  color: #aeb2bd;
}

.c0:disabled svg {
  fill: #aeb2bd;
}

<div>
  <button
    class="c0"
    data-testid="ion-chip"
  >
    <span>
      Ragnarok!
    </span>
  </button>
</div>
`;

exports[`IonChip Custom Props should render chip selected 1`] = `
.c0 {
  background-color: #fcfcfd;
  border: 1px solid #ced2db;
  border-radius: 100px;
  color: #0858ce;
  cursor: pointer;
  font-weight: 600;
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 8px;
  height: 24px;
  padding: 4px 8px;
  font-size: 1.2rem;
  line-height: 1.6rem;
  background-color: #ebf3fe;
  border: 1.5px solid #0858ce;
  color: #0858ce;
}

.c0:focus-visible {
  outline: 2px solid #146ff5;
  outline-offset: 2px;
}

.c0 svg {
  fill: #0858ce;
}

.c0:hover,
.c0:focus-visible {
  background-color: #ebf3fe;
  border-color: #84b4fa;
  color: #146ff5;
}

.c0:hover svg,
.c0:focus-visible svg {
  fill: #146ff5;
}

.c0:active {
  background-color: #b5d2fc;
  border-color: #146ff5;
  color: #06439d;
}

.c0:active svg {
  fill: #06439d;
}

.c0:disabled {
  cursor: not-allowed;
  background-color: #e4e6eb;
  border-color: #aeb2bd;
  color: #aeb2bd;
}

.c0:disabled svg {
  fill: #aeb2bd;
}

.c0 svg {
  fill: #0858ce;
}

.c0:hover {
  background-color: #b5d2fc;
  border-color: #146ff5;
  color: #146ff5;
}

.c0:hover svg {
  fill: #146ff5;
}

.c0:focus-visible {
  background-color: #ebf3fe;
  border-color: #5396f8;
  color: #146ff5;
}

.c0:focus-visible svg {
  fill: #146ff5;
}

.c0:active {
  background-color: #84b4fa;
  border-color: #06439d;
  color: #06439d;
}

.c0:active svg {
  fill: #06439d;
}

.c0:disabled {
  background-color: #e4e6eb;
  border-color: #aeb2bd;
  color: #aeb2bd;
}

.c0:disabled svg {
  fill: #aeb2bd;
}

<div>
  <button
    class="c0"
    data-testid="ion-chip"
  >
    <span>
      Ragnarok!
    </span>
  </button>
</div>
`;

exports[`IonChip Custom Props should render chip sm size 1`] = `
.c0 {
  background-color: #fcfcfd;
  border: 1px solid #ced2db;
  border-radius: 100px;
  color: #0858ce;
  cursor: pointer;
  font-weight: 600;
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 8px;
  height: 24px;
  padding: 4px 8px;
  font-size: 1.2rem;
  line-height: 1.6rem;
}

.c0:focus-visible {
  outline: 2px solid #146ff5;
  outline-offset: 2px;
}

.c0 svg {
  fill: #0858ce;
}

.c0:hover,
.c0:focus-visible {
  background-color: #ebf3fe;
  border-color: #84b4fa;
  color: #146ff5;
}

.c0:hover svg,
.c0:focus-visible svg {
  fill: #146ff5;
}

.c0:active {
  background-color: #b5d2fc;
  border-color: #146ff5;
  color: #06439d;
}

.c0:active svg {
  fill: #06439d;
}

.c0:disabled {
  cursor: not-allowed;
  background-color: #e4e6eb;
  border-color: #aeb2bd;
  color: #aeb2bd;
}

.c0:disabled svg {
  fill: #aeb2bd;
}

<div>
  <button
    class="c0"
    data-testid="ion-chip"
  >
    <span>
      Ragnarok!
    </span>
  </button>
</div>
`;
